From e6abd74c222f0899554ad0a78b5ddd6f8e1a0975 Mon Sep 17 00:00:00 2001
From: Konstantina Chremmou <konstantina.chremmou@citrix.com>
Date: Thu, 4 Jun 2020 23:56:54 +0100
Subject: [PATCH] Branding for the Stockholm release.

---
 ocaml/idl/datamodel_types.ml | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/ocaml/idl/datamodel_types.ml b/ocaml/idl/datamodel_types.ml
index 1622520a6..0270a82a7 100644
--- a/ocaml/idl/datamodel_types.ml
+++ b/ocaml/idl/datamodel_types.ml
@@ -266,8 +266,8 @@ let release_order_full = [{
      code_name     = Some rel_stockholm;
      version_major = 2;
      version_minor = 15;
-     branding      = "Unreleased";
-     release_date  = None;
+     branding      = "Citrix Hypervisor 8.2";
+     release_date  = Some "June 2020";
    }
   ]
 (* When you add a new release, use the version number of the latest release, "Unreleased"
-- 
2.24.1.windows.2

