From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrii Sultanov <andriy.sultanov@vates.tech>
Date: Fri, 6 Mar 2026 09:45:45 +0000
Subject: [PATCH] xapi.conf: Switch to optimized data cluster format for VHD
 and QCOW

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

diff --git a/scripts/xapi.conf b/scripts/xapi.conf
index c8e451d37703..cbc4cc3fcbe2 100644
--- a/scripts/xapi.conf
+++ b/scripts/xapi.conf
@@ -419,3 +419,6 @@ create-tools-sr = true
 
 # Allows to modify the host's scheduler granularity
 allow-host-sched-gran-modification = true
+
+# Switches to optimized data cluster format for VHD and QCOW
+vhd-legacy-blocks-format=false
