From 8f7d7593a73322dd55a5c9ffdd041644edb9aead Mon Sep 17 00:00:00 2001
From: Guillaume <guillaume.thouvenin@vates.tech>
Date: Thu, 3 Jul 2025 17:27:40 +0200
Subject: [PATCH] fixup! Allow selection of image format during VM migration

---
 ocaml/xapi/storage_migrate.ml | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/ocaml/xapi/storage_migrate.ml b/ocaml/xapi/storage_migrate.ml
index 246d44bb7..f070a499b 100644
--- a/ocaml/xapi/storage_migrate.ml
+++ b/ocaml/xapi/storage_migrate.ml
@@ -724,11 +724,12 @@ module MigrateLocal = struct
   let start ~task ~dbg ~sr ~vdi ~image_format ~dp ~mirror_vm ~copy_vm ~url ~dest
       ~verify_dest =
     SXM.info
-      "%s sr:%s vdi:%s dp: %s mirror_vm: %s copy_vm: %s url:%s dest:%s \
+      "%s sr:%s vdi:%s image_format: %s dp: %s mirror_vm: %s copy_vm: %s url:%s dest:%s \
        verify_dest:%B"
       __FUNCTION__
       (Storage_interface.Sr.string_of sr)
       (Storage_interface.Vdi.string_of vdi)
+      image_format
       dp
       (Storage_interface.Vm.string_of mirror_vm)
       (Storage_interface.Vm.string_of copy_vm)
@@ -872,6 +873,7 @@ module MigrateLocal = struct
         (string_of_vdi_info local_vdi) ;
       let local_vdi = add_to_sm_config local_vdi "mirror" ("nbd:" ^ dp) in
       let local_vdi = add_to_sm_config local_vdi "base_mirror" mirror_id in
+      let local_vdi = add_to_sm_config local_vdi "image-format" image_format in
       let snapshot =
         try Local.VDI.snapshot dbg sr local_vdi with
         | Storage_interface.Storage_error (Backend_error (code, _))
