From 47e0bb912f28e2b6bab686d0f38571d95af3c600 Mon Sep 17 00:00:00 2001
From: Guillaume <guillaume.thouvenin@vates.tech>
Date: Thu, 3 Jul 2025 17:27:40 +0200
Subject: [PATCH] fixup! Allow selection of image format during VM migration

---
 ocaml/xapi/storage_migrate.ml | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/ocaml/xapi/storage_migrate.ml b/ocaml/xapi/storage_migrate.ml
index 246d44bb7..886b9b8bd 100644
--- a/ocaml/xapi/storage_migrate.ml
+++ b/ocaml/xapi/storage_migrate.ml
@@ -724,11 +724,12 @@ module MigrateLocal = struct
   let start ~task ~dbg ~sr ~vdi ~image_format ~dp ~mirror_vm ~copy_vm ~url ~dest
       ~verify_dest =
     SXM.info
-      "%s sr:%s vdi:%s dp: %s mirror_vm: %s copy_vm: %s url:%s dest:%s \
+      "%s sr:%s vdi:%s image_format: %s dp: %s mirror_vm: %s copy_vm: %s url:%s dest:%s \
        verify_dest:%B"
       __FUNCTION__
       (Storage_interface.Sr.string_of sr)
       (Storage_interface.Vdi.string_of vdi)
+      image_format
       dp
       (Storage_interface.Vm.string_of mirror_vm)
       (Storage_interface.Vm.string_of copy_vm)
@@ -868,10 +869,11 @@ module MigrateLocal = struct
       debug "%s Updated mirror_id %s in the active local mirror" __FUNCTION__
         mirror_id ;
 
-      SXM.info "%s About to snapshot VDI = %s" __FUNCTION__
-        (string_of_vdi_info local_vdi) ;
       let local_vdi = add_to_sm_config local_vdi "mirror" ("nbd:" ^ dp) in
       let local_vdi = add_to_sm_config local_vdi "base_mirror" mirror_id in
+      let local_vdi = add_to_sm_config local_vdi "image-format" image_format in
+      SXM.info "%s About to snapshot VDI = %s" __FUNCTION__
+        (string_of_vdi_info local_vdi) ;
       let snapshot =
         try Local.VDI.snapshot dbg sr local_vdi with
         | Storage_interface.Storage_error (Backend_error (code, _))
