diff --git i/ocaml/idl/datamodel.ml w/ocaml/idl/datamodel.ml
index f2ea73999..f78293b79 100644
--- i/ocaml/idl/datamodel.ml
+++ w/ocaml/idl/datamodel.ml
@@ -6519,11 +6519,39 @@ end
 (** PCI devices *)

 module PCI = struct
+  let hide =
+    call ~name:"hide" ~lifecycle:[(Published, rel_vgpu_tech_preview, "")]
+    ~versioned_params:
+      [
+        {
+          param_type= Ref _pci
+        ; param_name= "self"
+        ; param_doc= "The PCI to hide"
+        ; param_release= next_release
+        ; param_default= None
+        }
+      ]
+    ~allowed_roles:_R_POOL_OP ()
+
+  let unhide =
+    call ~name:"unhide" ~lifecycle:[(Published, rel_vgpu_tech_preview, "")]
+    ~versioned_params:
+      [
+        {
+          param_type= Ref _pci
+        ; param_name= "self"
+        ; param_doc= "The PCI to unhide"
+        ; param_release= next_release
+        ; param_default= None
+        }
+      ]
+    ~allowed_roles:_R_POOL_OP ()
+
   let t =
     create_obj ~name:_pci ~descr:"A PCI device" ~doccomments:[]
       ~gen_constructor_destructor:false ~gen_events:true ~in_db:true
       ~lifecycle:[(Published, rel_boston, "")]
-      ~messages:[] ~messages_default_allowed_roles:_R_POOL_OP
+      ~messages:[hide; unhide] ~messages_default_allowed_roles:_R_POOL_OP
       ~persist:PersistEverything ~in_oss_since:None ~db_logging:Log_destroy
       ~contents:
         [
diff --git i/ocaml/xapi/message_forwarding.ml w/ocaml/xapi/message_forwarding.ml
index 4d08bb593..0fa61321c 100644
--- i/ocaml/xapi/message_forwarding.ml
+++ w/ocaml/xapi/message_forwarding.ml
@@ -5841,7 +5841,23 @@ functor

     module Secret = Local.Secret

-    module PCI = struct end
+    module PCI = struct
+      let hide ~__context ~self =
+        info "PCI.hide: pci = '%s'" (pci_uuid ~__context self) ;
+        let host = Db.PCI.get_host ~__context ~self in
+        let local_fn = Local.PCI.hide ~self in
+        do_op_on ~__context ~local_fn ~host (fun session_id rpc ->
+            Client.PCI.hide ~rpc ~session_id ~self
+        )
+
+      let unhide ~__context ~self =
+        info "PCI.unhide: pci = '%s'" (pci_uuid ~__context self) ;
+        let host = Db.PCI.get_host ~__context ~self in
+        let local_fn = Local.PCI.unhide ~self in
+        do_op_on ~__context ~local_fn ~host (fun session_id rpc ->
+            Client.PCI.unhide ~rpc ~session_id ~self
+      )
+    end

     module VTPM = struct
       let create ~__context ~vM ~is_unique =
diff --git i/ocaml/xapi/xapi_pci.ml w/ocaml/xapi/xapi_pci.ml
index 6e72c366e..2ab955e4c 100644
--- i/ocaml/xapi/xapi_pci.ml
+++ w/ocaml/xapi/xapi_pci.ml
@@ -319,3 +319,9 @@ let get_system_display_device () =
       )
       None items
   with _ -> None
+
+let hide ~__context ~self =
+  Pciops.hide_pci ~__context self
+
+let unhide ~__context  ~self =
+  Pciops.unhide_pci ~__context self
diff --git i/ocaml/xapi/xapi_pci.mli w/ocaml/xapi/xapi_pci.mli
index dd71dfffc..bc937c8c3 100644
--- i/ocaml/xapi/xapi_pci.mli
+++ w/ocaml/xapi/xapi_pci.mli
@@ -51,3 +51,9 @@ val disable_system_display_device : unit -> unit

 val dequarantine : __context:Context.t -> Xenops_interface.Pci.address -> unit
 (** dequarantine a PCI device. This is idempotent. *)
+
+val hide : __context:Context.t -> self:API.ref_PCI -> unit
+(** hide a PCI device. *)
+
+val unhide : __context:Context.t -> self:API.ref_PCI -> unit
+(** unhide a PCI device. *)
