diff --git a/ocaml/xapi/xapi_tunnel.ml b/ocaml/xapi/xapi_tunnel.ml
index b94fd38b22..37f1baf1ec 100644
--- a/ocaml/xapi/xapi_tunnel.ml
+++ b/ocaml/xapi/xapi_tunnel.ml
@@ -44,6 +44,9 @@ let create_internal ~__context ~transport_PIF ~network ~host ~protocol =
   let device_name = device in
   let mAC = Xapi_vif_helpers.gen_mac (0, Uuid.to_string (Uuid.make_uuid ())) in
   let metrics = Db.PIF.get_metrics ~__context ~self:transport_PIF in
+  let primary_address_type =
+    Db.PIF.get_primary_address_type ~__context ~self:transport_PIF
+  in
   Db.PIF.create ~__context ~ref:access_PIF
     ~uuid:(Uuid.to_string (Uuid.make_uuid ()))
     ~device ~device_name ~network ~host ~mAC ~mTU:(-1L) ~vLAN:(-1L) ~metrics
@@ -51,7 +54,7 @@ let create_internal ~__context ~transport_PIF ~network ~host ~protocol =
     ~ip_configuration_mode:`None ~iP:"" ~netmask:"" ~gateway:"" ~dNS:""
     ~bond_slave_of:Ref.null ~vLAN_master_of:Ref.null ~management:false
     ~other_config:[] ~disallow_unplug:false ~ipv6_configuration_mode:`None
-    ~iPv6:[""] ~ipv6_gateway:"" ~primary_address_type:`IPv4 ~managed:true
+    ~iPv6:[""] ~ipv6_gateway:"" ~primary_address_type ~managed:true
     ~properties:[] ~capabilities:[] ~pCI:Ref.null ;
   Db.Tunnel.create ~__context ~ref:tunnel
     ~uuid:(Uuid.to_string (Uuid.make_uuid ()))
