---
diff --git a/scripts/xapi.conf b/scripts/xapi.conf
index 797aff1a3..615a488b1 100644
--- a/scripts/xapi.conf
+++ b/scripts/xapi.conf
@@ -1,3 +1,8 @@
+# WARNING: any changes made to this file may be overwritten by future
+# updates of the xapi-core RPM.
+# You may want to create a file in the /etc/xapi.conf.d/ directory instead
+
+
 # The xapi config file has 3 sections: a global policy section;
 # a section for paths to helper utilities; and a timeout tweak section.
 
@@ -153,7 +158,7 @@ sparse_dd = /usr/libexec/xapi/sparse_dd
 # sm-dir =  @OPTDIR@/sm
 
 # Whitelist of SM plugins
-sm-plugins=ext nfs iscsi lvmoiscsi dummy file hba rawhba udev iso lvm lvmohba lvmofcoe shm smb
+sm-plugins=ext nfs iscsi lvmoiscsi dummy file hba rawhba udev iso lvm lvmohba lvmofcoe shm smb cephfs glusterfs xfs zfs ext4
 
 # Directory containing tools ISO
 # tools-sr-dir = @OPTDIR@/packages/iso
@@ -377,3 +382,6 @@ xen_livepatch_list = "/usr/sbin/xen-livepatch list"
 
 # The command to query current kernel patch list
 kpatch_list = "/usr/sbin/kpatch list"
+
+# true = forbid HTTP requests to website (403); false = allow HTTP to HTTPS redirection
+website-https-only = false
-- 
2.40.1

