diff --git c/scripts/xapi.conf w/scripts/xapi.conf
index 46f859a8d..ddbf923f3 100644
--- c/scripts/xapi.conf
+++ w/scripts/xapi.conf
@@ -1,3 +1,7 @@
+# WARNING: any changes made to this file may be overwritten by future
+# updates of the xapi-core RPM.
+# You may want to create a file in the /etc/xapi.conf.d/ directory instead
+
 # The xapi config file has 3 sections: a global policy section;
 # a section for paths to helper utilities; and a timeout tweak section.

@@ -82,7 +86,7 @@ igd-passthru-vendor-whitelist = 8086
 # repository-domain-name-allowlist =

 # Allow the use of custom UEFI certificates
-# allow-custom-uefi-certs = false
+allow-custom-uefi-certs = true

 # Paths to utilities: ############################################

@@ -163,7 +167,7 @@ sparse_dd = /usr/libexec/xapi/sparse_dd
 # sm-dir =  @OPTDIR@/sm

 # Whitelist of SM plugins
-sm-plugins=ext nfs iscsi lvmoiscsi dummy file hba rawhba udev iso lvm lvmohba lvmofcoe shm smb
+sm-plugins=ext nfs iscsi lvmoiscsi dummy file hba rawhba udev iso lvm lvmohba lvmofcoe shm smb cephfs glusterfs xfs zfs largeblock

 # Directory containing tools ISO
 # tools-sr-dir = @OPTDIR@/packages/iso
@@ -396,3 +400,12 @@ xen_livepatch_list = "/usr/sbin/xen-livepatch list"

 # The command to query current kernel patch list
 kpatch_list = "/usr/sbin/kpatch list"
+
+# true = forbid HTTP requests to website (403); false = allow HTTP to HTTPS redirection
+website-https-only = false
+
+# Indicates whether to create an SR for Tools ISOs
+create-tools-sr = true
+
+# Allows to modify the host's scheduler granularity
+allow-host-sched-gran-modification = true
