diff --git i/ocaml/quicktest/qt_filter.ml w/ocaml/quicktest/qt_filter.ml
index 6d11a3252..91fcd7221 100644
--- a/ocaml/quicktest/qt_filter.ml
+++ b/ocaml/quicktest/qt_filter.ml
@@ -170,15 +170,15 @@ module SR = struct
 
   let all_srs = lazy (list_srs_connected_to_localhost !A.rpc !session_id)
 
-  let all =
+  let all () =
     if !A.use_default_sr then
       let pool = Qt.get_pool !A.rpc !session_id in
       only
         (Client.Client.Pool.get_default_SR ~rpc:!A.rpc ~session_id:!session_id
            ~self:pool
-        )
+        ) ()
     else
-      fun () -> Lazy.force all_srs
+      Lazy.force all_srs
 
   let random srs () =
     let srs = srs () in
-- 
2.40.1

