diff --git a/ocaml/xapi/xapi_vlan.ml b/ocaml/xapi/xapi_vlan.ml
index 074951300f..dbab38e40c 100644
--- a/ocaml/xapi/xapi_vlan.ml
+++ b/ocaml/xapi/xapi_vlan.ml
@@ -35,6 +35,9 @@ let create_internal ~__context ~host ~tagged_PIF ~tag ~network ~device =
   (* Copy the MTU and metrics from the base PIF *)
   let mTU = Db.PIF.get_MTU ~__context ~self:tagged_PIF in
   let metrics = Db.PIF.get_metrics ~__context ~self:tagged_PIF in
+  let primary_address_type =
+    Db.PIF.get_primary_address_type ~__context ~self:tagged_PIF
+  in
   Db.PIF.create ~__context ~ref:untagged_PIF
     ~uuid:(Uuid.to_string (Uuid.make_uuid ()))
     ~device ~device_name:device ~network ~host ~mAC:vlan_mac ~mTU ~vLAN:tag
@@ -43,7 +46,7 @@ let create_internal ~__context ~host ~tagged_PIF ~tag ~network ~device =
     ~netmask:"" ~gateway:"" ~dNS:"" ~bond_slave_of:Ref.null ~vLAN_master_of:vlan
     ~management:false ~other_config:[] ~disallow_unplug:false
     ~ipv6_configuration_mode:`None ~iPv6:[""] ~ipv6_gateway:""
-    ~primary_address_type:`IPv4 ~managed:true ~properties:[] ~capabilities:[]
+    ~primary_address_type ~managed:true ~properties:[] ~capabilities:[]
     ~pCI:Ref.null ;
   let () =
     Db.VLAN.create ~__context ~ref:vlan ~uuid:vlan_uuid ~tagged_PIF
