See: https://github.com/xapi-project/xen-api/pull/5068

diff --git a/ocaml/xapi/xapi_host.ml b/ocaml/xapi/xapi_host.ml
index c83886f9d7..18d26a189d 100644
--- a/ocaml/xapi/xapi_host.ml
+++ b/ocaml/xapi/xapi_host.ml
@@ -2527,7 +2527,12 @@ let migrate_receive ~__context ~host ~network ~options:_ =
         let configuration_mode =
           Db.PIF.get_ipv6_configuration_mode ~__context ~self:pif
         in
-        match Db.PIF.get_IPv6 ~__context ~self:pif with
+        let valid_ipv6s =
+          List.filter
+            (fun ipv6 -> not (String.starts_with ~prefix:"fe80::" ipv6))
+            (Db.PIF.get_IPv6 ~__context ~self:pif)
+        in
+        match valid_ipv6s with
         | [] ->
             ("", configuration_mode)
         | ip :: _ ->
diff --git a/ocaml/xapi/xapi_pif_helpers.ml b/ocaml/xapi/xapi_pif_helpers.ml
index 29c91cc805..2a9ebf9000 100644
--- a/ocaml/xapi/xapi_pif_helpers.ml
+++ b/ocaml/xapi/xapi_pif_helpers.ml
@@ -262,4 +262,9 @@ let get_primary_address ~__context ~pif =
     match Db.PIF.get_IP ~__context ~self:pif with "" -> None | ip -> Some ip
   )
   | `IPv6 ->
-      List.nth_opt (Db.PIF.get_IPv6 ~__context ~self:pif) 0
+      let valid_ipv6s =
+        List.filter
+          (fun ipv6 -> not (String.starts_with ~prefix:"fe80::" ipv6))
+          (Db.PIF.get_IPv6 ~__context ~self:pif)
+      in
+      List.nth_opt valid_ipv6s 0
