diff --git a/ocaml/libs/xapi-stdext/lib/xapi-fd-test/test/test_xapi_fd_test.ml b/ocaml/libs/xapi-stdext/lib/xapi-fd-test/test/test_xapi_fd_test.ml
index b6ae12eb0..77de7cdf2 100644
--- a/ocaml/libs/xapi-stdext/lib/xapi-fd-test/test/test_xapi_fd_test.ml
+++ b/ocaml/libs/xapi-stdext/lib/xapi-fd-test/test/test_xapi_fd_test.ml
@@ -106,10 +106,4 @@ let () =
     [
       ("test_kind_ro", test_kind_all test_kind_ro)
     ; ("test_kind_wo", test_kind_all test_kind_wo)
-    ; ( "cancellable sleep"
-      , [
-          Alcotest.test_case "cancellable" `Quick test_cancellable_sleep
-        ; Alcotest.test_case "full" `Quick test_full_sleep
-        ]
-      )
     ]
