--- xapi-24.19.2/ocaml/xenopsd/scripts/qemu-wrapper.orig	2024-10-29 14:48:40.032356048 +0000
+++ xapi-24.19.2/ocaml/xenopsd/scripts/qemu-wrapper	2024-10-29 14:50:04.456362180 +0000
@@ -293,7 +293,7 @@
 
     print("Exec: %s %s" % (qemu_dm, " ".join(qemu_args)))
 
-    clipboardd = '/opt/xensource/libexec/xs-clipboardd'
+    clipboardd = '/opt/xensource/libexec/xcp-clipboardd'
     subprocess.call([clipboardd, "-d", str(domid), "-s", str(s2.fileno())],
                     pass_fds=[s2.fileno()])
 
